textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/join/SimilarityJoinPredicate.java

```diff
@@ -0,0 +1,195 @@
+package edu.uci.ics.textdb.dataflow.join;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import edu.uci.ics.textdb.api.common.Attribute;
+import edu.uci.ics.textdb.api.common.FieldType;
+import edu.uci.ics.textdb.api.common.IField;
+import edu.uci.ics.textdb.api.common.ITuple;
+import edu.uci.ics.textdb.api.common.Schema;
+import edu.uci.ics.textdb.common.constants.SchemaConstants;
+import edu.uci.ics.textdb.common.exception.DataFlowException;
+import edu.uci.ics.textdb.common.field.DataTuple;
+import edu.uci.ics.textdb.common.field.IDField;
+import edu.uci.ics.textdb.common.field.ListField;
+import edu.uci.ics.textdb.common.field.Span;
+import edu.uci.ics.textdb.dataflow.common.IJoinPredicate;
+import info.debatty.java.stringsimilarity.NormalizedLevenshtein;
+
+public class SimilarityJoinPredicate implements IJoinPredicate {
+    
+    public static final String INNER_PREFIX = "inner_";
+    public static final String OUTER_PREFIX = "outer_";
+    
+    Double similarity;
+    String joinAttributeName;
+    
+    /**
+     * The output schema of Similarity Join will be the combination of inner and outer schema.
+     *   (except the _id, spanList and payload field)
+     *   
+     * Example of a same-table, different-tuple join
+     * table_schema,   inner_tuple,             outer_tuple
+     *   _id:          random_id                random_id
+     *   content:      "join"                   "john"
+     *   spanList:     (0, 4, "join", content)  (0, 4, "john", content)
+     *   payload:      payload_inner            payload_outer
+     *   
+     * result_schema,      result_tuple
+     *   _id:              new_random_id
+     *   inner_content:    "join"
+     *   outer_content:    "john"
+     *   spanList:         [(0, 4, "join", inner_content)  (0, 4, "john", outer_content)]
+     *   payload:          [payload_inner, payload_outer]
+     *   
+     */
+    public SimilarityJoinPredicate(String joinAttributeName, Double similarity) {
+        if (similarity > 1) {
+            similarity = 1.0;
+        } else if (similarity < 0) {
+            similarity = 0.0;
+        }
+        this.similarity = similarity;
+        this.joinAttributeName = joinAttributeName;
+    }
+
+    @Override
+    public String getIDAttributeName() {
+        return SchemaConstants._ID;
+    }
+
+    @Override
+    public String getJoinAttributeName() {
+        return joinAttributeName;
+    }
+    
+    @Override
+    public Schema generateOutputSchema(Schema innerOperatorSchema, Schema outerOperatorSchema) throws DataFlowException {
```

Comment:
In case this function is called multiple times with the same result, do you think it's better to compute it once then "cache" it by storing it into a variable?